src/pg_query_summary_statement_type.c

```diff
@@ -0,0 +1,798 @@
+#include "pg_query.h"
+#include "pg_query_internal.h"
+#include "pg_query_outfuncs.h"
+
+#include "parser/parser.h"
+#include "parser/scanner.h"
+#include "parser/scansup.h"
+
+#include "nodes/nodeFuncs.h"
+#include "utils/builtins.h"
+
+#include <unistd.h>
+#include <fcntl.h>
+#include <string.h>
+
+#include "pg_query_summary.h"
+
+// Declare the huge function early so we can put it at the end of the file.
+bool pg_query_summary_stmt_walk_impl(Node* node, Summary* summary);
+
+// Adds new_stmt_type to summary->statement_types, if it's not already in it.
+// What we really want is a set, but I'd rather not implement one in C.
+void add_stmt_type_if_needed(Summary* summary, char* new_stmt_type) {
+	ListCell* lc = NULL;
+	foreach(lc, summary->statement_types) {
+		char* stmt_type = lfirst(lc);
+		if (strcmp(new_stmt_type, stmt_type) == 0)
+			return;
+	}
+
+	summary->statement_types = lappend(summary->statement_types, new_stmt_type);
+}
+
+
+PgQueryError* pg_query_summary_statement_walk(Summary* summary, Node* tree) {
+	pg_query_summary_stmt_walk_impl(tree, summary);
+
+	return NULL;
+}
+
+bool pg_query_summary_stmt_walk_impl(Node* node, Summary* summary) {
+	if (node == NULL) return false;
+
+	switch(nodeTag(node)) {
+		case T_AlterCollationStmt:
+			add_stmt_type_if_needed(summary, "AlterCollationStmt");
```

Comment:
I think all of the utility statements here (i.e. things that are not SELECT/UPDATE/DELETE/INSERT/MERGE) are not handled by the tree walker, so we should return early. You can check https://github.com/postgres/postgres/blob/master/src/backend/nodes/nodeFuncs.c#L4000 to confirm.